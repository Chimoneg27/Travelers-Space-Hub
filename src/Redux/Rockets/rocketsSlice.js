/* eslint-disable import/no-extraneous-dependencies */
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const rocketAPI = 'https://api.spacexdata.com/v4/rockets';

export const getRocketData = createAsyncThunk('rockets/getRocketData', async () => {
  const response = await axios.get(rocketAPI);
  return response.data;
});

const rocketReducer = createSlice({
  name: 'SpaceX Rockets',
  initialState: {
    rocketArr: [],
    isLoading: false,
    error: null,
    isDataFetched: false,
  },
  reducers: {
    addReservation: (state, action) => {
      const updatedRocketArr = state.rocketArr.map((rocketObj) => (rocketObj.id === action.payload
        ? { ...rocketObj, reserved: false } : rocketObj));
      return { ...state, rocketArr: updatedRocketArr };
    },
    cancelReservation: (state, action) => {
      const updatedRocketArr = state.rocketArr.map((rocketObj) => (rocketObj.id === action.payload
        ? { ...rocketObj, reserved: true } : rocketObj));
      return { ...state, rocketArr: updatedRocketArr };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getRocketData.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(getRocketData.fulfilled, (state, action) => {
        state.isLoading = false;
        state.rocketArr = action.payload.map((rocket) => ({
          id: rocket.id,
          name: rocket.name,
          flickr_images: rocket.flickr_images,
          description: rocket.description,
        }));
        state.isDataFetched = true;
      })
      .addCase(getRocketData.rejected, (state, action) => {
        state.isLoading = false;
        state.error = action.error;
      });
  },
});

export const { cancelReservation, addReservation } = rocketReducer.actions;
export default rocketReducer.reducer;
