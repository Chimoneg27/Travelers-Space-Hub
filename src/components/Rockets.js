import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getRocketData } from '../Redux/Rockets/rocketsSlice';
import RocketRender from './RocketRender';
import '../Styles/RocketsRender.css';

const Rockets = () => {
  const rockets = useSelector((state) => state.rockets.rocketArr);
  const isLoading = useSelector((state) => state.rockets.isLoading);
  const dataFetched = useSelector((state) => state.rockets.isDataFetched);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!dataFetched) {
      dispatch(getRocketData());
    }
  }, [dispatch, dataFetched]);

  return (
    <div className="rocket-page">
      {isLoading ? (
        <h2>Loading...</h2>
      ) : (
        <div className="rocket-container">
          {rockets.map((rocket) => (
            <RocketRender
              key={rocket.id}
              rocket={rocket}
            />
          ))}
        </div>
      )}
    </div>
  );
};

export default Rockets;
