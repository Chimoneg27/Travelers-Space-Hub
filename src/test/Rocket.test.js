/* eslint-disable import/no-extraneous-dependencies */
import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Rockets from '../components/Rockets';

const mockStore = configureMockStore([]);

test('renders rockets component correctly', async () => {
  const store = mockStore({
    rockets: {
      rocketArr: [
        {
          id: '1',
          name: 'Rocket1',
          flickr_images: ['image1'],
          description: 'Description1',
        },
        {
          id: '2',
          name: 'Rocket2',
          flickr_images: ['image2'],
          description: 'Description2',
        },

      ],
      isDataFetched: true,
    },
  });
  const RocketsComponent = () => (
    <Provider store={store}>
      <Rockets />
    </Provider>
  );
  const { getText } = render(<RocketsComponent />);
  expect(getText).toMatchSnapshot();
  expect(screen.getByText('Rocket1')).toBeInTheDocument();
  expect(screen.getByText('Rocket2')).toBeInTheDocument();
  expect(screen.getByText('Description1')).toBeInTheDocument();
});
