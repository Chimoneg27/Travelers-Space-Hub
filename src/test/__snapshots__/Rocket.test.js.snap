// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders rockets component correctly 1`] = `undefined`;
